tree_segment.py

```diff
@@ -0,0 +1,434 @@
+import os
+
+
+class Point_Cloud:
+    # object that stores the points in a point cloud, initialized from a las file
+    # points can be enabled or disabled, disabled points are ignored when searching for stems
+    def __init__(self, file_name, verbose=False):
+        all_data = laspy.read(file_name)
+        height_of_points_above_ground = np.array(
+            [point[-1] for point in all_data.points.array])
+        self.points = [Point(xyz, height) for xyz, height in zip(
+            all_data.xyz, height_of_points_above_ground)]
+        # internal, to access this use enabled_points()
+        self.__enabled_points__ = self.points
+        self.verbose = verbose
+
+    def enabled_points(self):
+        # filters the set of points for just those enabled, then returns that set of enabled points
+        self.__enabled_points__ = filter_for_enabled(self.__enabled_points__)
+        return self.__enabled_points__
+
+    def find_highest_point(self):
+        # returns the highest point in the point cloud, ie the one with the greatest z coordinate
+        return max(self.enabled_points(), key=lambda point: point.z())
+
+    def find_stem(self, min_height=.5, fail_to_find_disable_radius=.2, descendents_height=.3):
+        # returns a stem from the point cloud if it exists, otherwise returns None
+        # the stem will be a list of points in descending z value, starting from above the min height and going down to a ground point
+        # inputs:
+        #   - min_height : the minimum height (in m) for a valid stem
+        #   - fail_to_find_disable_radius : when a point has no descendents, we disable each point within this radius of it
+        #   - descendents_height : the maximum vertical distance allowed to a descendent of this point
+        stem_grounded = False
+        while not(stem_grounded):
+            # the stem starts at the highest point in the cloud
+            stem = [self.find_highest_point()]
+            if stem[0].height_above_ground < min_height:
+                # if your stem is too short, we're done
+                break
+            while not(stem_grounded) and stem:
+                # we try to find a point in a cylinder below this stem
+                active_point = stem[-1]
+                descendents = active_point.find_descendents(
+                    self.enabled_points(), height_below_self=descendents_height)
+                if descendents:
+                    # if there is a valid descendent, take one at random (we can try again later if it doesnt work)
+                    next_point = random.choice(descendents)
+                    stem.append(next_point)
+                    if next_point.is_ground:
+                        stem_grounded = True
+                else:
+                    # if the point has no descendent, it cannot have a path to the ground, so we disable it and all points within fail_to_find_disable_radius of it
+                    n_points_start = len(self.enabled_points())
```

Comment:
Rather than tracking and printing the number of active points here. Consider moving this logic into `disable_stem_region`. e.g. distable_stem_region could return the number of newly disabled points and/or could log that value.

Reply:
Done